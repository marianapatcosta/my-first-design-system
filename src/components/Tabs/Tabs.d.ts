import React, { CSSProperties } from 'react';
export interface TabsProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
    /**
     * for accessibility purposes indicates for which purpose will Tabs be used?
     */
    tabsPurpose?: string;
    /**
     * content for Tabs
     */
    tabsMetadata: Array<{
        label: string;
        content: JSX.Element;
    }>;
    /**
     * Are Tabs disabled?
     */
    disabled?: boolean;
}
export declare const Tabs: React.FC<TabsProps>;
