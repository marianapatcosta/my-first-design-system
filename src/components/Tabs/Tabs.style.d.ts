export declare const StyledTabs: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledTabsListWrapper: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledTabsList: import("styled-components").StyledComponent<"ul", any, {}, never>;
export declare const StyledTabsContent: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledTabsSpacer: import("styled-components").StyledComponent<"div", any, {}, never>;
