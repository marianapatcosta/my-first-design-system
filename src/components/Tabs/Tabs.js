var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from 'react';
import { Tab } from '../Tab/Tab';
import { StyledTabs, StyledTabsListWrapper, StyledTabsList, StyledTabsContent, StyledTabsSpacer, } from './Tabs.style';
export var Tabs = function (_a) {
    var disabled = _a.disabled, tabsPurpose = _a.tabsPurpose, tabsMetadata = _a.tabsMetadata, className = _a.className, props = __rest(_a, ["disabled", "tabsPurpose", "tabsMetadata", "className"]);
    var _b = useState(+localStorage.getItem('activeTab') || 0), activeTabIndex = _b[0], setActiveTabIndex = _b[1];
    var onClickTab = function (tabIndex) {
        setActiveTabIndex(tabIndex);
        localStorage.setItem('activeTab', '' + tabIndex);
    };
    return (_jsxs(StyledTabs, __assign({ className: className }, props, { children: [_jsx(StyledTabsListWrapper, { children: _jsx(StyledTabsList, __assign({ role: 'tablist', "aria-label": "Select " + tabsPurpose }, { children: tabsMetadata.map(function (_a, index) {
                        var label = _a.label;
                        return (_jsx(Tab, { label: label, disabled: disabled, isActive: index === activeTabIndex, onClick: function () { return onClickTab(index); } }, "tab-" + (index + Math.random())));
                    }) }), void 0) }, void 0),
            _jsxs(StyledTabsContent, __assign({ role: 'tabpanel' }, { children: [_jsx(StyledTabsSpacer, {}, void 0), tabsMetadata[activeTabIndex].content] }), void 0)] }), void 0));
};
Tabs.defaultProps = {
    id: '',
    className: '',
    style: {},
    tabsPurpose: '',
    tabsMetadata: [],
    disabled: false,
};
