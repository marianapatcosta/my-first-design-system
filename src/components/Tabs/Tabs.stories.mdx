import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { styled } from '@storybook/theming';
import { Tabs } from "./Tabs";

export const StyledDiv = styled.div`
  & .fEpyfw {
    position: relative;
  }
`

<Meta title="Design System/Tabs" component={Tabs} />

# Tabs
<br/><br/> 

### Display different content divided among Tabs's Tabs components

Use Tabs when different contents should be shown in separate, according to the user's needs; each 
of these contents is displayed in a Tab component

<Preview withToolbar>
  <Story name="Standard">
    <StyledDiv>
      <Tabs
        tabsMetadata={[
          { label: "Tab 1", content: <div>Tab 1 content</div> },
          { label: "Tab 2", content: <div>Tab 2 content</div> },
        ]}
      />
    </StyledDiv>
  </Story>
</Preview>

<ArgsTable of={Tabs} />

### Playground

Experiment with Tabs Component with the Controls addon in the Canvas tab.

export const Template = (args) => <Tabs {...args} />

<Canvas>
    <Story name="Controls" args={{
    tabsMetadata: [
        { label: 'Tab 1', content: <div>Tab 1 content</div>  },
        { label: 'Tab 2', content: <div>Tab 2 content</div> },
      ]
    }}>
      {Template.bind({})}
    </Story>
</Canvas> 