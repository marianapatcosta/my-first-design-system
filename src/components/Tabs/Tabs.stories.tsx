import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Tabs, TabsProps } from './Tabs';

export default {
  title: 'Tabs',
  component: Tabs,
  // args behavior in each story
  args: {
    tabsMetadata: [
      { label: 'Tab 1', content: <div>Tab 1 content</div>  },
      { label: 'Tab 2', content: <div>Tab 2 content</div> },
    ],
  },
} as Meta;

const Template: Story<TabsProps> = (args) => <Tabs {...args} />;

export const Default = Template.bind({});
Default.args = {};
