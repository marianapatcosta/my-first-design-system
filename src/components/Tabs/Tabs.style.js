var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
export var StyledTabs = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), baseStyles);
export var StyledTabsListWrapper = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: fixed;\n  z-index: 1;\n  background-color: ", ";\n  width: 100%;\n  height: 4rem;\n"], ["\n  position: fixed;\n  z-index: 1;\n  background-color: ", ";\n  width: 100%;\n  height: 4rem;\n"])), colors.primary);
export var StyledTabsList = styled.ul(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  align-items: flex-end;\n  justify-content: flex-start;\n  list-style: none;\n  margin: 1.25rem 0;\n  padding: 0.625rem 1.25rem 0;\n  box-shadow: 0 0.25rem 0.5rem -0.5rem ", ";\n"], ["\n  display: flex;\n  align-items: flex-end;\n  justify-content: flex-start;\n  list-style: none;\n  margin: 1.25rem 0;\n  padding: 0.625rem 1.25rem 0;\n  box-shadow: 0 0.25rem 0.5rem -0.5rem ", ";\n"])), colors.shadow);
export var StyledTabsContent = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject([""], [""])));
export var StyledTabsSpacer = styled.div(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  height: 4rem;\n  margin-bottom: 1rem;\n"], ["\n  height: 4rem;\n  margin-bottom: 1rem;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
