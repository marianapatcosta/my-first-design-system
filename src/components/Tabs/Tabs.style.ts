import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';

export const StyledTabs = styled.div`
  ${baseStyles}
`;

export const StyledTabsListWrapper = styled.div`
  position: fixed;
  z-index: 1;
  background-color: ${colors.primary};
  width: 100%;
  height: 4rem;
`;

export const StyledTabsList = styled.ul`
  display: flex;
  align-items: flex-end;
  justify-content: flex-start;
  list-style: none;
  margin: 1.25rem 0;
  padding: 0.625rem 1.25rem 0;
  box-shadow: 0 0.25rem 0.5rem -0.5rem ${colors.shadow};
`;

export const StyledTabsContent = styled.div``;

export const StyledTabsSpacer = styled.div`
  height: 4rem;
  margin-bottom: 1rem;
`;
