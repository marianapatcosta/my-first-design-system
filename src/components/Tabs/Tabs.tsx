import React, { CSSProperties, useState } from 'react';
import { Tab } from '../Tab/Tab';
import {
  StyledTabs,
  StyledTabsListWrapper,
  StyledTabsList,
  StyledTabsContent,
  StyledTabsSpacer,
} from './Tabs.style';

export interface TabsProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
  /**
   * for accessibility purposes indicates for which purpose will Tabs be used?
   */
  tabsPurpose?: string;
  /**
   * content for Tabs
   */
  tabsMetadata: Array<{ label: string; content: JSX.Element }>;
  /**
   * Are Tabs disabled?
   */
  disabled?: boolean;
}

export const Tabs: React.FC<TabsProps> = ({
  disabled,
  tabsPurpose,
  tabsMetadata,
  className,
  ...props
}) => {
  const [activeTabIndex, setActiveTabIndex] = useState<number>(
    +localStorage.getItem('activeTab')! || 0
  );

  const onClickTab = (tabIndex: number): void => {
    setActiveTabIndex(tabIndex);
    localStorage.setItem('activeTab', '' + tabIndex);
  };

  return (
    <StyledTabs className={className} {...props}>
      <StyledTabsListWrapper>
        <StyledTabsList role='tablist' aria-label={`Select ${tabsPurpose}`}>
          {tabsMetadata.map(({ label }, index: number) => (
            <Tab
              key={`tab-${index + Math.random()}`}
              label={label}
              disabled={disabled}
              isActive={index === activeTabIndex}
              onClick={() => onClickTab(index)}
            />
          ))}
        </StyledTabsList>
      </StyledTabsListWrapper>
      <StyledTabsContent
        role='tabpanel'
        aria-labelledby={tabsMetadata[activeTabIndex].label}
      >
        <StyledTabsSpacer />
        {tabsMetadata[activeTabIndex].content}
      </StyledTabsContent>
    </StyledTabs>
  );
};

Tabs.defaultProps = {
  id: '',
  className: '',
  style: {},
  tabsPurpose: '',
  tabsMetadata: [],
  disabled: false,
};
