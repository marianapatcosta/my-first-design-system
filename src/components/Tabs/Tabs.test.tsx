import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Tabs } from './Tabs';
import { Tab } from '../Tab/Tab';

describe('Tabs', () => {
  const defaultProps = {
    tabsMetadata: [
      { label: 'Tab 1', content: <div>Tab 1 content</div> },
      { label: 'Tab 2', content: <div>Tab 2 content</div> },
    ],
  };

  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Tabs {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display Tab 1 content by default', () => {
    const { queryByText } = render(<Tabs {...defaultProps} />);
    expect(queryByText('Tab 1 content')).toBeTruthy();
    expect(queryByText('Tab 2 content')).toBeNull();
  });

  test('should display Tab 2 content when Tab 2 is clicked', () => {
    const { queryByText } = render(<Tabs {...defaultProps} />);

    const tab2 = queryByText('Tab 2');
    tab2 && fireEvent.click(tab2);

    expect(queryByText('Tab 2 content')).toBeTruthy();
    expect(queryByText('Tab 1 content')).toBeNull();
  });
});
