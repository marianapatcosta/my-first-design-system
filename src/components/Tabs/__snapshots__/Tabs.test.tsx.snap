// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders without errors and matches snapshot 1`] = `
<div
  class="sc-gsTCUz bGRYEi"
  id=""
>
  <div
    class="sc-dlfnbm cHVBSC"
  >
    <ul
      aria-label="Select "
      class="sc-hKgILt dHFEmC"
      role="tablist"
    >
      <li
        aria-selected="true"
        class="sc-bdfBwQ csIuzl"
        id=""
        role="tab"
        tabindex="0"
      >
        Tab 1
      </li>
      <li
        aria-selected="false"
        class="sc-bdfBwQ dElXjk"
        id=""
        role="tab"
        tabindex="0"
      >
        Tab 2
      </li>
    </ul>
  </div>
  <div
    aria-labelledby="Tab 1"
    class="sc-eCssSg"
    role="tabpanel"
  >
    <div
      class="sc-jSgupP wnDln"
    />
    <div>
      Tab 1 content
    </div>
  </div>
</div>
`;
